Reticulate support

dfalbel opened the pull request:
Add support for reticulate sessions. See https://github.com/posit-dev/positron/pull/4603

lionel- commented on crates/ark/src/reticulate.rs:
```diff
@@ -0,0 +1,145 @@
+        };
+
+        let event = CommManagerEvent::Opened(service.comm.clone(), serde_json::Value::Null);
+        unwrap!(service.comm_manager_tx.send(event), Err(e) => {
```
If return value is not used, better use the standard `if let Err(err)`

dfalbel replied:
Used `.or_log_error()` instead

lionel- commented on crates/ark/src/reticulate.rs:
```diff
@@ -0,0 +1,145 @@
+        }
+    );
+
+    if let Some(id) = (*comm_id_guard).clone() {
```
Nit: More idiomatic and transparent to clone at the call site that needs to own the object (i.e. innermost scope).

lionel- commented on crates/ark/src/reticulate.rs:
```diff
@@ -0,0 +1,145 @@
+        return Ok(R_NilValue);
+    }
+
+    let input = RObject::try_from(input)?;
```
We're converging towards this style which makes for clearer code thanks to the type annotation:

```suggestion
    let input: RObject = input.try_into()?;
```

lionel- replied:
Also for consistency with the line below.

lionel- commented on crates/ark/src/reticulate.rs:
```diff
@@ -0,0 +1,145 @@
+        }
+
+        let mut comm_id_guard = unwrap!(
+            RETICULATE_COMM_ID.try_lock(),
```
`try_lock()` is very low level. Note that it's allowed to fail spuriously. Should use `.lock().unwrap()` here, as we are not worried about contention or deadlocks.

lionel- commented on crates/ark/src/modules/positron/reticulate.R:
```diff
@@ -0,0 +1,99 @@
+#' Used by the front-end to install reticulate
+#' A modal asking to install is shown before calling this.
+#' @export
+.ps.rpc.install_reticulate <- function() {
```
Actually shouldn't the frontend use the existing way of installing packages, i.e. `r.packageInstall()`?

dfalbel replied:
`r.packageInstall` is essentially just `devtools::install()` it doesn't allow installing arbitrary packages.
I can add `.ps.rpc.install_package(pkg)` if we want a more general approach. 

yes, the front-end can set verbosity

lionel- commented on crates/ark/src/modules/positron/reticulate.R:
```diff
@@ -0,0 +1,99 @@
+.ps.rpc.reticulate_start_kernel <- function(kernelPath, connectionFile, logFile, logLevel) {
+    tryCatch({
+        reticulate:::py_run_file_on_thread(
+            file = kernelPath,
+            args = c(
+                "-f", connectionFile,
```
I assume this is for running ipykernel in the Python thread with the connection information generated by Positron? I think that could use some comments about what is going on.

dfalbel replied:
Added some comments! It would be nice to keep this close to the call site: https://github.com/posit-dev/positron/blob/5f3a1fa6c9d9e0eeae5b63483024e852fc5acc83/extensions/positron-reticulate/src/extension.ts#L383-L390 but `RuntimeSession.execute` does not allow awaiting /and or handling errors.

lionel- commented on crates/ark/src/reticulate.rs:
```diff
@@ -0,0 +1,145 @@
+                break;
+            }
+
+            // Forward data msgs to the frontend
```
We're just resending the data messages back? Could you add comments about that please?

dfalbel replied:
We don't need at, I copied from another comm by mistake.

lionel- commented on crates/ark/src/reticulate.rs:
```diff
@@ -0,0 +1,145 @@
+pub unsafe extern "C" fn ps_reticulate_open(input: SEXP) -> Result<SEXP, anyhow::Error> {
+    let main = RMain::get();
+
+    if !RMain::initialized() {
+        return Ok(R_NilValue);
+    }
```
This should probably surface an error.

But once https://github.com/posit-dev/ark/pull/542 is merged, you can use `RMain::wait_initialized()`. If you need R to be ready too, you can also use `RMain::wait_r_initialized()`.

That said this method is called from R so actually this probably should be removed altogether.

dfalbel replied:
This was required for testing the connections comm, because loading a package could call comm_open, but it's not required here and I accidentally copied it over.

lionel- commented:
My main comment is about `try_lock()` which should be replaced by `lock()` to avoid rarely occurring surprising behaviour.

dfalbel commented:
I addressed most points in https://github.com/posit-dev/ark/pull/506/commits/286a24aa0000ba736694dd133e9de117ba9d7d20

Also moved `.ps.rpc.install_reticulate()` into a more general `.ps.rpc.install_packages`.

lionel- reviewed (approved):
LG!
